ponix4-src/source/acpid-2.0.23/kacpimon/kacpimon.8
.TH kacpimon 8 "Jan 2013"
.\" Copyright (c) 2009-2013 Ted Felix (www.tedfelix.com)
.SH NAME
kacpimon \- Kernel ACPI Event Monitor
.SH SYNOPSIS
\fBkacpimon\fP
.SH DESCRIPTION
\fBkacpimon\fP is a monitor program that connects to three sources of ACPI 
events:
.IP
1. /proc/acpi/event
.br
2. netlink
.br
3. the input layer (/dev/input/event*)
.PP
\fBkacpimon\fP then reports on what it sees while it is connected.
.PP
Use \fBkacpimon\fP to make sure your kernel is sending acpi-related events
via the three sources mentioned above.
.PP
Since both \fBacpid\fP(8) and \fBkacpimon\fP connect to the same sources, be sure
to stop \fBacpid\fP(8) before running kacpimon.  If you are using 
\fBupstart\fP(7), use \fBinitctl\fP(8) to stop \fBacpid\fP(8).  If you are using 
\fBsystemd\fP(1), use \fBsystemctl\fP(1) to stop \fBacpid\fP(8).  If you aren't
using either of those, you can try \fBkillall\fP(1) to stop \fBacpid\fP(8).
.PP
Be sure to run \fBkacpimon\fP as root or else it won't be able to open
/proc/acpi/event or the input layer:
.IP
sudo kacpimon
.PP  
If you want to pipe the output of \fBkacpimon\fP to a file, be sure to use
the "Esc" key to exit rather than Ctrl-C.  Using Ctrl-C will cause
output to be lost.
.SH FILES
.TP
.I /proc/acpi/event
kernel ACPI event file
.TP
.I /dev/input/event*
input layer event files
.SH SEE ALSO
.BR acpi_listen (8),
.BR acpid (8)
.SH AUTHOR
Ted Felix <www.tedfelix.com>
.SH TODO
\- Make kacpimon something that actually gets installed?
.SH LICENSE
kacpimon \- Kernel ACPI Event Monitor
.br
Copyright (C) 2009-2013  Ted Felix (www.tedfelix.com)
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
.PP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
.PP
You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
Or visit www.gnu.org.



ponix4-const/history.txt
2011-02-03 19:45 - 2.00.575-beta - added graphical libs to usb-ip package
2011-02-04 14:45 - 2.00.576-beta -
2011-02-07 18:10 - 2.00.577-beta - build MTRR (vesa not working)
2011-02-07 18:19 - 2.00.578-beta -
2011-02-10 20:02 - 2.00.579-beta - libsane 1.0.21 with avision backend from git
2011-02-25 18:09 - 2.00.580-beta - edit libsane package
2011-02-25 18:16 - 2.00.581-beta -
2011-02-27 19:50 - 2.00.582-beta - usb-server package first build. included like package
2011-02-28 21:00 - 2.00.583-beta - final release
2011-02-28 21:07 - 2.00.584 - final release
2011-02-28 21:21 - 2.00.585 - see previos message
2011-02-28 21:45 - 2.1.001 - new version format
2011-02-28 22:51 - 2.1.1 - new version
2011-02-28 23:38 - 2.1.2 - sane package
2011-03-01 20:50 - 2.1.3 - new package sane 1.0.22
2011-03-01 23:05 - 2.1.4 -
2011-03-31 22:06 - 2.1.5 - lzma for package and usb2com converter
2011-04-18 16:44 - 2.1.6 - rxvt.pkg
2011-04-18 18:33 - 2.1.7 - rxvt.pkg ver2
2011-04-18 18:41 - 2.1.8 - rxvt ver3
2011-05-06 15:25 - 2.1.9 - repair freerdp server connect
2011-06-29 21:06 - 2.1.10 - old ver
2011-06-29 23:52 - 2.1.11 - intel video driver 2.15.0, freerdp from git, package rdesktop 1.7
2011-06-30 00:06 - 2.1.12 -
2011-06-30 00:34 - 2.1.13 -
2011-06-30 13:27 - 2.1.14 -
2011-06-30 14:20 - 2.1.15 -
2011-06-30 15:14 - 2.1.16 -
2011-06-30 15:40 - 2.1.17 -
2011-06-30 16:04 - 2.1.18 -
2011-07-01 00:11 - 2.1.19 -
2011-07-01 01:15 - 2.1.20 -
2011-07-05 20:51 - 2.1.21-standalone - rename to standalone version
2011-07-06 14:16 - 2.1.22 - add intel driver 2.15.0 and last version libdrm
2011-07-06 18:37 - 2.1.23 - rdesktop+keymaps
2011-07-07 16:05 - 2.1.24 - add keymaps to the freerdp17 package
2011-07-07 19:16 - 2.1.25 - error in the rdesktop package
2011-07-07 19:50 - 2.1.26 - too
2011-07-07 20:10 - 2.1.27 - vmview auto start
2011-07-07 20:16 - 2.1.28 - too
2011-07-07 21:08 - 2.1.29 - edit vmopenclient
2011-07-08 00:11 - 2.1.30 - edit vmopenclient
2011-07-12 21:18 - 2.1.31 - etoken
2011-07-13 18:28 - 2.1.32 - repair sane package
2011-07-13 19:02 - 2.1.33 - add smart-card support to rdesktop17 package
2011-07-21 13:38 - 2.1.34 - last version of realtek ethernet driver r8101
2011-07-21 14:52 - 2.1.35 - add firmware fo rtl8105e
2011-07-21 19:18 - 2.1.36 - new version of realtek r8169 driver (15)
2011-07-21 20:55 - 2.1.37 - without r8101.ko
2011-07-21 21:27 - 2.1.38 - r8169 from kernek, r8101 delete
2011-07-22 16:45 - 2.1.39 - add citrix client package
2011-07-22 16:51 - 2.1.40 -
2011-07-22 19:03 - 2.1.41 -
2011-07-22 19:59 - 2.1.42 -
2011-07-22 20:07 - 2.1.43 -
2011-07-22 20:40 - 2.1.44 -
2011-07-23 14:11 - 2.1.45-standalone - trying wifi build with odinary /etc/init.d/network (no ethtool)
2011-07-26 13:16 - 2.1.46 -
2011-07-26 15:20 - 2.1.47 -
2011-07-26 16:22 - 2.1.48 -
2011-07-27 14:30 - 2.1.49 -
2011-07-27 15:43 - 2.1.50 -
2011-07-27 18:05 - 2.1.51 -
2011-08-01 13:59 - 2.1.52 -
2011-08-01 15:36 - 2.1.53 -
2011-08-01 16:25 - 2.1.54 -
2011-08-01 18:18 - 2.1.55 -
2011-08-01 20:09 - 2.1.56 -
2011-08-01 20:10 - 2.1.57 - renew xorg server and libXfont
2011-08-01 20:43 - 2.1.58 -
2011-08-01 21:40 - 2.1.59 -
2011-08-01 22:32 - 2.1.60 -
2011-08-01 23:24 - 2.1.61 -
2011-08-01 23:37 - 2.1.62 -
2011-08-02 13:53 - 2.1.63 -
2011-08-02 15:07 - 2.1.64 -
2011-08-02 19:31 - 2.1.65 -
2011-08-02 20:30 - 2.1.66 -
2011-08-02 21:16 - 2.1.67 -
2011-08-02 22:29 - 2.1.68 -
2011-08-02 22:37 - 2.1.69 -
2011-08-03 16:08 - 2.1.70 -
2011-08-03 19:12 - 2.1.71 -
2011-08-03 19:47 - 2.1.72 - ICA 11 packages working!
2011-08-03 20:34 - 2.1.73 -
2011-08-03 20:43 - 2.1.74 -
2011-08-10 22:32 - 2.1.75 - add no_tls and no_nla to freerdp parameters
2011-08-11 00:30 - 2.1.76-standalone -
2011-08-11 18:35 - 2.1.77 - repair bug with freerdp_no_nla freerdp_no_tls
2011-08-11 21:45 - 2.1.78 - repair bug with Xorg (absent libfreetype)
2011-08-12 15:15 - 2.1.79 - add relatime parameter to udev script
2011-08-16 15:07 - 2.1.80 -
2011-08-16 17:26 - 2.1.81 -
2011-08-16 17:51 - 2.1.82 -
2011-08-16 21:43 - 2.1.83 - add gstreamer for ica11 package
2011-08-16 22:31 - 2.1.84 - add gstreamer and plugins for ica11 package
2011-08-18 13:57 - 2.1.85 - add gstream for speedscreen function for citrix client
2011-08-18 14:47 - 2.1.86 - add oss support to the alsa drivers
2011-08-18 15:11 - 2.1.87 -
2011-08-18 16:08 - 2.1.88 - rebuild libasound
2011-08-18 18:01 - 2.1.89 - error, soundcore must be from the kernel
2011-08-18 18:55 - 2.1.90 -
2011-09-12 18:06 - 2.1.91 - correct error with rdesktop17 package add pcsclite library
2011-09-12 22:31 - 2.1.92 - add icewm
2011-09-12 23:00 - 2.1.93 - add graphical libs to icewm
2011-09-12 23:11 - 2.1.94 - add locales to icewm
2011-09-12 23:37 - 2.1.95 -
2011-09-12 23:45 - 2.1.96 -
2011-09-13 00:07 - 2.1.97 -
2011-09-13 00:22 - 2.1.98 -
2011-09-13 13:32 - 2.1.99 -
2011-09-13 14:17 - 2.1.100 - 1
2011-09-13 19:30 - 2.1.101 -
2011-09-13 20:01 - 2.1.102 -
2011-09-13 20:40 - 2.1.103 -
2011-09-13 23:41 - 2.1.104 -
2011-09-14 15:00 - 2.1.105 -
2011-09-14 15:37 - 2.1.106 -
2011-09-16 15:08 - 2.1.107 -
2011-09-26 21:38 - 2.1.108 - in network_script added SERVER_IP=$serverid. hack to use dhcp server as tftp when no siaddr specified (current udhcpc has a bug and doesn't set tftp option)
2011-09-27 20:49 - 2.1.109 - last version of nvidia xorg driver
2011-09-28 14:48 - 2.1.110 - add --no-nla parameter for freerdp client
2011-10-03 20:00 - 2.1.111 - new version of atheros ethernet driver atl1e (AR81Family-linux-1.0.1.14) instead atl1c and atl1e (now removed)
2011-10-03 20:24 - 2.1.112-standalone - current update
2011-10-03 20:52 - 2.1.113 -
2011-10-03 21:22 - 2.1.114 -
2011-10-03 22:10 - 2.1.115 - udev-debug
2011-10-03 22:56 - 2.1.116 - udev debug
2011-10-03 23:23 - 2.1.117 - normal udev
2011-10-03 23:36 - 2.1.118 - udev debug
2011-10-03 23:42 - 2.1.119 - udev debug
2011-10-04 14:48 - 2.1.120 - corrected bug with network_script
2011-10-04 18:28 - 2.1.121 - add atl1c
2011-10-04 18:41 - 2.1.122 -
2011-10-04 19:32 - 2.1.123 - rollback to 1.0.0.9 atl1e driver

ponix4-src/source/acpid-2.0.23/kacpimon/kacpimon.c
/*
 *  kacpimon - Kernel ACPI Event Monitor
 *
 *  Monitors kernel ACPI events from multiple interfaces and reports them
 *  to the console.
 *
 *  Inspired by (and in some cases blatantly lifted from) Vojtech Pavlik's
 *  evtest.c, Zhang Rui's acpi_genl, and Alexey Kuznetsov's libnetlink.
 *
 *  Copyright (C) 2008, Ted Felix (www.tedfelix.com)
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 *  (tabs at 4)
 */

/* system */
#include <stdio.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/select.h>
#include <string.h>
#include <errno.h>
#include <malloc.h>

/* local */
#include "libnetlink.h"
#include "genetlink.h"
#include "acpi_genetlink.h"

#include "acpi_ids.h"
#include "connection_list.h"
#include "input_layer.h"
#include "netlink.h"

#include "kacpimon.h"

/* ??? Isn't this in a system header someplace? */
#define max(a, b)  (((a)>(b))?(a):(b))

/*********************************************************************/

/* Exit flag that can be set by any of the functions to cause the */
/* program to exit.  */
int exitflag = 0;

/****************************************************************
 *  Old /proc/acpi/event interface
 ****************************************************************/

static void process_proc(int fd)
{
	const int buffsize = 1024;
	char buffer[buffsize];
	ssize_t nbytes;

	for (;;)
	{
		nbytes = read(fd, buffer, buffsize - 1);

		/* ??? Do we need to worry about partial messages? */
		
		/* If there are no data to read, bail. */
		if (nbytes <= 0)
			break;

		/* Ensure we have a zero terminator */
		buffer[nbytes] = 0;
		
		printf("/proc/acpi/event: %s", buffer);
	}
}

// ---------------------------------------------------------------

static void open_proc(void)
{
	char *filename = "/proc/acpi/event";
	int fd;
	struct connection c;

	fd = open(filename, O_RDONLY | O_NONBLOCK);
	if (fd >= 0)
	{
		printf("%s opened successfully\n", filename);

		/* Add a connection to the list. */
		c.fd = fd;
		c.process = process_proc;
		add_connection(&c);
	}
	else
	{
		int errno2 = errno;
		printf("open for %s: %s (%d)\n", 
			filename, strerror(errno2), errno2);
		if (errno2 == EACCES)
			printf("  (try running as root)\n");
		if (errno2 == ENOENT)
			printf("  (ACPI proc filesystem may not be present)\n");
		if (errno2 == EBUSY)
			printf("  (ACPI proc filesystem is in use.  "
				"You might need to kill acpid.)\n");
	}
}

/****************************************************************
 *  Main Program Functions
 ****************************************************************/

static void monitor(void)
{
	while (exitflag == 0)
	{
		fd_set readfds;
		int nready;
		int i;

		/* It's going to get clobbered, so use a copy. */
		readfds = *get_fdset();

		/* Wait on data. */
		nready = select(get_highestfd() + 1, &readfds, NULL, NULL, NULL);

		/* If something goes wrong, bail. */
		if (nready <= 0)
			break;

		/* For each connection */
		for (i = 0; i <= get_number_of_connections(); ++i)
		{
			int fd;
			struct connection *p;

			p = get_connection(i);

			/* If this connection is invalid, bail. */
			if (!p)
				break;

			/* Get the file descriptor */
			fd = p -> fd;

			/* If this file descriptor has data waiting, */
			if (FD_ISSET(fd, &readfds))
			{
				p->process(fd);
			}
		}
	}
}

// ---------------------------------------------------------------

static void close_all(void)
{
	int i = 0;

	/* For each connection */
	for (i = 0; i <= get_number_of_connections(); ++i)
	{
		struct connection *p;

		p = get_connection(i);

		/* If this connection is invalid, try the next. */
		if (p == 0)
			continue;

		close(p -> fd);
	}
}

// ---------------------------------------------------------------

int main(void)
{
	printf("Kernel ACPI Event Monitor...\n");

	open_proc();
	open_input();
	open_netlink();

	printf("Press Escape to exit, or Ctrl-C if that doesn't work.\n");

	monitor();

	printf("Closing files...\n");

	close_all();

	printf("Goodbye\n");

	return 0;
}
